File: 1905-count-sub-islands/1905-count-sub-islands.py
class Solution(object):
    def countSubIslands(self, grid1, grid2):
        """        
        :type grid1: List[List[int]]
        :type grid2: List[List[int]]
        :rtype: int
        """
        row, col = len(grid2), len(grid2[0])
        visited = [[0] * col for _ in range(row)]
        directions = [(0, 1), (0, -1), (1, 0), (-1, 0)]

        def bfs(start_r, start_c):
            queue = deque([(start_r, start_c)])
            is_sub_island = True
            
            while queue:
                cur_r, cur_c = queue.popleft()
                for dr, dc in directions:
                    nr, nc = cur_r + dr, cur_c + dc
                    if 0 <= nr < row and 0 <= nc < col and grid2[nr][nc] == 1 and not visited[nr][nc]:
                        visited[nr][nc] = 1
                        queue.append((nr, nc))
                        if grid1[nr][nc] == 0:
                            is_sub_island = False
            return is_sub_island

        cnt = 0
        for r in range(row):
            for c in range(col):
                if not visited[r][c] and grid2[r][c] == 1 and grid1[r][c] == 1:
                    visited[r][c] = 1
                    if bfs(r, c):
                        cnt += 1
                        
        return cnt

File: 1905-count-sub-islands/1905-count-sub-islands.cpp
#include <iostream>
#include <vector>
#include <queue>
#define X first
#define Y second
using namespace std;

class Solution {
public:
    int dx[4] = {1, 0, -1, 0};
    int dy[4] = {0, 1, 0, -1};
    bool bfs(queue<pair<int, int>>& Q, vector<vector<int>>& visited, vector<vector<int>>& grid1, vector<vector<int>>& grid2) {
        bool is_sub = true;
        while (!Q.empty()){
            pair<int, int> cur = Q.front(); Q.pop();
            for (int dir=0; dir < 4; dir++){
                int nx = cur.X + dx[dir];
                int ny = cur.Y + dy[dir];
                if (nx < 0 || ny < 0 || nx >= grid2.size() || ny >= grid2[0].size()) continue;
                if (grid2[nx][ny] == 1 && visited[nx][ny] == 0) {
                    Q.push({nx, ny});
                    visited[nx][ny] =1;
                    if (grid1[nx][ny] == 0) is_sub = false;
                }

            }
        }
        return is_sub;
    }

    int countSubIslands(vector<vector<int>>& grid1, vector<vector<int>>& grid2) {
        int row = grid2.size(); int col = grid2[0].size();
        vector<vector<int>> visited(row, vector<int>(col, 0));
        queue<pair<int, int>> Q;
        int cnt=0;

        for (int i=0; i < row;  i++){
            for (int j=0; j < col; j++){
                if (!visited[i][j] && grid2[i][j] == 1 && grid1[i][j] == 1) {
                    visited[i][j] = 1;
                    Q.push({i, j});
                    if (bfs(Q, visited, grid1, grid2)) {
                        cnt++;
                    }
                }
            }
        }
        return cnt; 
    }
};